---
title: "AMMS Population Structure from Allele Frequencies Practical" 
author: "Izzy Routledge, Bob Verity"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    theme: readable
    highlight: tango
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: false
    toc_depth: 2
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F, 
                      fig.align = 'center', fig.keep = 'all')
```

<!-- Here we style out button a little bit -->

```{=html}
<style>
.showopt {
background-color: #004c93;
color: #FFFFFF; 
width: 100px;
height: 20px;
text-align: center;
vertical-align: middle !important;
border-radius: 8px;
float:right;
}

.showopt:hover {
background-color: #dfe4f2;
color: #004c93;
}

</style>
```

# Dependencies for Practical {-}

Please copy and paste the below code chunk in it's entirety to your console to download R package libraries needed for this practical. If you are having trouble installing any of the R packages, please ask an instructor for a pre-loaded flash drive.

```{r, echo=T}
deps <- c("tidyverse", "genescaper", "MIPanalyzer","rmaverick")
deps <- !sapply(deps, function(x){x %in% installed.packages()[,1]})
if (any(deps)) {
  if (deps["tidyverse"]) {
    install.packages("tidyverse")
  }
  if (deps["genescaper"]) {
    devtools::install_github("mrc-ide/genescaper", ref = "v0.1.0")
  }
  if (deps["MIPanalyzer"]) {
    devtools::install_github("mrc-ide/MIPanalyzer", ref = "v1.0.0")
  }
  if (deps["rmaverick"]) {
    devtools::install_github("bobverity/rmaverick", ref = "v1.1.0")
  }
}

```

Now load all of those libraries into this session using the code chunk below. Please copy and paste it in its entirety.

```{r, echo=T}
library(tidyverse)
library(genescaper)
library(MIPanalyzer)
library(rmaverick)
```

Finally, source the additional functions that are needed for this practical by copy-pasting this function:

```{r, echo=T}
source("source_functions/pop_structure_utils.R")
```


# Intro to population structure and allele frequency metrics {-}


One of the most fundamental observations that we can make from genetic data is how the frequencies of different alleles vary over time and over space.

For loci where mutations confer drug resistance, allele frequencies tell us something with direct clinical implications - if an allele is at high frequencies then we might consider switching first-line drugs.

However, even for loci that have no direct impact on phenotype, and therefore are selectively neutral, allele frequencies can give us important information about what is going on in the parasite population. 

These loci will be influenced by factors such as prevalence, human and mosquito population size, and migration rates between different sub-populations. If we can correctly interpret these signals then we can use neutral allele frequencies to tell us things about population structure and connectivity that can be relevant for control purposes.

_**Definition:** An allele is the identity of a particular genetic locus or sequence that is inherited between parents and offspring._

_**Definition:** A locus is a fixed position on a chromosome where a particular genetic marker is located._

_**Definition:** A neutral allele is one which is not positively or negatively affecting an organism's fitness and undergoing selection._

### Overview of Data {-}

We will work with both simulated and real data in this practical. The simulated data will allow us to explore how allele frequencies change over time as a function of different underlying parameters - something that we can almost never do from messy real-world data. This will allow us to build up an intuition about what we would expect to see in real data. In the second half of the practical, we will use real-world datasetes including a large molecular inversion probe (MIP) dataset taken from DRC and surrounding countries, previously analysed by [Verity et al. 2020](https://pubmed.ncbi.nlm.nih.gov/32355199/). We will use this to explore population structure in various ways.

### Practical Goals {-}
By the end of this practical, you should be able to: 

- Describe how population size, mutation and migration rates affect the patterns that we expect to see in neutral allele frequency data
- Estimate metrics such as $F_{ST}$ and explain what they can tell us about population differentiation
- Use ordination techniques like PCA to explore patterns of population structure
- Use model-based techniques like *rmaverick* to detect population structure in a probability-driven way
- Appreciate how allele frequency methods can identify connectivity over large spatial scales and long temporal scales
- Explain how allele frequencies can give useful information, and also describe their limitations for malaria control purposes


-------------------------

# Population genetics through simulation {-}
In this section, we will simulate various populations with different frequencies of alleles be using the R function: `sim_freqs`. Note, this function was loaded from `source_functions/pop_structure_utils.R`. 

## Single subpopulation {-}

Let's first consider a **single** subpopulation (e.g. village) where there is no barrier to gene flow, also known as a "deme". As a result, a deme is a group of individuals (*e.g.* parasites) that can freely interbreed. 

### Allele frequency changes in a single deme {-}

First we will simulate the allele frequencies of a population of 1000 individuals infected with malaria. 

```{r}
# Note, participants should not include the quiet() function - run: sim_freqs(N = 1000, mut_rate = 0)
sim1 <- quiet(sim_freqs(N = 1000, mut_rate = 0))

```
_**Question/Coding Challenge**_: Try running this function lots of times. What do you notice about the allele frequency over time?

_**Question/Coding Challenge**_: Now repeat with a smaller population size (which represents the number of *infected* hosts in our case); for example, `N=500`, `N=100`, and `N=50`. What do you notice about the strength of drift? What happens once an allele reaches a frequency of 0 or 1? Why is this?

```{r}
# Note, participants should not include the quiet() function 
quiet(sim_freqs(N = 100, mut_rate = 0))
```

_**Definition:** Drift is the variation in the frequency of alleles in a small population due to the random loss of alleles across generations._

Next, we will explore how our simulation of a single, larger subpopulation changes when adding mutations and considering the population for some time. 

_**Question/Coding Challenge**_: Now repeat the simulation, this time running for 10 years and with some mutation. Do frequencies still always end up fixed/lost?

```{r}
# Note, participants should not include the quiet() function 
quiet(sim_freqs(N = 1000, t_out = seq(0, 365 * 10), mut_rate = 1e-4))
```

_**Question/Coding Challenge**_: Now repeat the simulation, but alter the mutation rate to be higher and lower respectively. How do the results change from above? (Example options are provided in the chunks below)
```{r, eval=F}
sim_freqs(N = 1000, t_out = seq(0, 365 * 10), mut_rate = 1e-2)
```

```{r, eval=F}
sim_freqs(N = 1000, t_out = seq(0, 365 * 10), mut_rate = 1e-10)
```

### The allele frequency distribution {-}
We will now simulate a large number of loci, but output will just consider a single timepoint: 5 years from the origin of our single subpopulation. Note, we will be using a moderate mutation rate as compared to the ones we considered above. 

```{r}
# simulate allele frequencies and look at the data
sim1 <- quiet(sim_freqs(N = 1000, t_out = 365 * 5, mut_rate = 1e-6, loci = 1e3, plot_on = FALSE))
head(sim1$data)
```
<br>
We can plot this allele frequency spectrum.  

_**Coding Challenge**_: Using the simulated data located in the `sim1$data` dataframe, make a histogram  with ggplot and the `geom_histogram` layer of the allele frequencies distributions. You can add breaks in the histogram using the following code: `breaks = seq(0, 1, 0.01)`.   

```{r}
# simulate allele frequencies and look at the data
ggplot(sim1$data) +
  geom_histogram(aes(x = freq), breaks = seq(0, 1, 0.01))
```

_**Question/Coding Challenge**_: What happens if you increase the mutation rate? What happens if you increase the population size? Example code chunks are provided below. 
```{r}
sim2 <- sim_freqs(N = 1e5, t_out = 365 * 5, mut_rate = 1e-6, loci = 1e3, plot_on = FALSE)
```
```{r}
sim3 <- sim_freqs(N = 1000, t_out = 365 * 5, mut_rate = 1e-3, loci = 1e3, plot_on = FALSE)
```
_**Coding Challenge**_: Can you write a function that pipes the simulation output directly into code that makes similar histogram plots above?  

<br>
<br>

### The buildup of homozygosity {-}
Now we will analyze how homozygosity changes over time in the setting of drift. 
_**Definition:** Homozygosity is the existence of identical allleles at a given locus._
<br>
<br>
_**Question/Coding Challenge**_: Run the code below multiple times to repeat the simulation of drift, but now also plotting homozygosity. What do you notice about how homozygosity changes over time?
<br>

```{r}
sim_freqs(N = 1000, t_out = seq(0, 365 * 10, 7), plot_homo = TRUE)
```
_**Answer**_: It always increases to 1
<br>
<br>

_**Question/Coding Challenge**_: Run the above code but now consider what happens when you add in some mutation? Try mutation rates of 1e-3 and 1e-4. What happens to the equilibrium level of homozygosity?

```{r}
sim_freqs(N = 1000, t_out = seq(0, 365 * 10, 7), mut_rate = 1e-4, plot_homo = TRUE)
```


```{r}
sim_freqs(N = 1000, t_out = seq(0, 365 * 5, 7), mut_rate = 1e-4, plot_homo = TRUE, plot_expected_homo = TRUE)
```
<br>
_**Question**_: What similarities do you notice between allele frequencies and homozygosity patterns?
<br>
<br>
<br>

## Multiple subpopulations {-}
In the prior section, you analyzed how allele frequencies changed under various scenarios in a single subpopulation. In this section, we will explore how these patterns change when multiple subpopulations (demes) are interacting (or not interacting). 

### Perfectly isolated subpopulations {-}
First, we will explore a scenario where we have 10 independent demes. This could represent ten villages where there is no movement of people (or parasites) between those villages.

_**Question/Coding Challenge**_: Run the simulation with 10 independent demes. What do you expect to happen to do the allele frequencies? Is the pattern for the indepedent demes similar or different than the pattern you observed when only considering a single subpopulation? 


```{r}
sim1 <- sim_freqs(N = 1000, demes = 10, mig_rate = 0, mut_rate = 1e-4)
```
<br>
_**Answer**_: In the independent populations, you can appreciate that the allele frequencies tend to diverge. This is the same behavior as our single subpopulation simulations above, but now replicated in 10 independent "trials" (or populations). 
<br>
<br>

#### Calculating Population Differentiation{-}
Next we will explore metrics used to measure population differentiation and contrast populations. We will focus on two different metrics of genetic differentiation among different subpopulations: the *fixation index*, or $F_{ST}$ and *Jost's D*.
<br>
High values of $F_{ST}$ and Jost's D both suggest a greater genetic differentiation between the subpopulations being measured. However both metrics have different approaches to measure genetic differentiation. 

<br>
<br>

##### Fst
We will first analyze $F_{ST}$. Note, when $F_{ST}$ is close to **1** it means that individuals in a population tend to carry a _small number of alleles_, or a _smaller number of segregating sites_. This is to say, within the population, individuals have genomes that are identical at numerous loci. As a result, individuals are “differentiated” in the sense that alleles are distinct between the populations and are not freely mixing between individuals between populations (*i.e.* they are segregated). 
<br>
Below, we have provided a simple version of $F_{ST}$ that can be calculated manually: 
$$
F_{ST} = \frac{\sigma^2_{\mbox{subpop}}}{\sigma^2_{\mbox{total}}} = \frac{\sigma^2_{\mbox{subpop}}}{\bar p (1-\bar p)}
$$
This equation says that $F_{ST}$ is a measure of the variance of allele frequencies among the subpopulations divided by the total variation in the allele frequency in the population, based on the defintion of binomial variance (as we are dealing with biallelic loci).

```{r}
sim_fst <- sim1$data %>%
  select(-locus) %>%
  group_by(time) %>%
  summarise(p_mean = mean(freq),
            p_var = var(freq),
            Fst = p_var / (p_mean * (1 - p_mean)))
```
<br>
<br>


_**Question/Coding Challenge**_: Now use the `geom_line` layer from ggplot to plot $F_{ST}$ over time (x-axisL time, y-axis: Fst). What do you notice about differentiation?

```{r}
ggplot(sim_fst) +
  geom_line(aes(x = time, y = Fst)) +
  expand_limits(y = c(0, 1))
```

_**Question/Coding Challenge**_: Now repeat the process above (simulation and plotting) with a higher mutation rate (`mut_rate = 1e-3`)? What changes do you notice? 

```{r}
sim1 <- sim_freqs(N = 100, demes = 10, mig_rate = 0, mut_rate = 1e-3,
                  t_out = seq(0, 365*10, 7), plot_on = FALSE)
p1 <- sim1$data %>%
  select(-locus) %>%
  group_by(time) %>%
  summarise(p_mean = mean(freq),
            p_var = popvar(freq),
            Fst = p_var / (p_mean * (1 - p_mean))) %>%
  ggplot() + theme_bw() +
  geom_line(aes(x = time, y = Fst)) +
  expand_limits(y = c(0, 1))
```

The equilibrium level of $F_{ST}$, where the mutation rate is in stead state with the rate of drift, is known to be $\frac{1}{(1 + 2N\mu)}$ in this case, where $N$ is the population size, and $\mu$ is the mutation rate. 

_**Question/Coding Challenge**_: Add this level in the plot above `geom_hline(yintercept = 1 / (1 + 2*1e2*1e-3), linetype = "dashed")`. How does this line compare with what you see in the plot?

```{r, eval = F}
p1 <- sim1$data %>%
  select(-locus) %>%
  group_by(time) %>%
  summarise(p_mean = mean(freq),
            p_var = popvar(freq),
            Fst = p_var / (p_mean * (1 - p_mean))) %>%
  ggplot() + theme_bw() +
  geom_line(aes(x = time, y = Fst)) +
  expand_limits(y = c(0, 1)) + 
  geom_hline(yintercept = 1 / (1 + 2*1e2*1e-3), linetype = "dashed")
```

The fact that $F_{ST}$ depends on the mutation rate can lead to problems. If one marker has a high mutation rate (e.g. microsatellites) then we would expect to see low values of $F_{ST}$, even if the populations have been evolving separately for a long time, and in this sense are well differentiated.

##### Jost's D
Jost's D is an alternative way of looking at differentiation that is less sensitive to the mutation rate.  

_**Coding Challenge**_: Simulate a population with 100 individuals, 10 demes, no migration, and a mutation rate of 1e-3. Consider timepoints every week for 10 years. 

```{r}
sim1 <- sim_freqs(N = 100, demes = 10, mig_rate = 0, mut_rate = 1e-3,
                  t_out = seq(0, 365*10, 7), plot_on = FALSE)
```


Now, we will calculate Jost's D, using the follow equation: 
$$
D = \frac{H_{Te}-H_{Se}}{1-H_{Se}} \frac{k}{k-1} 
$$
Jost's D is intuitively similar to Wright's $F_{ST}$, when $F_{ST}$ is considered under the alternative definition based on identity by state/descent (to be discussed later). As a result, the equation above states that the expected heterozygosity ($H_{.e}$) in the total population ($T$) can be contrasted with the expected heterozygosity in the subpopulation ($S$) to determine the overall population differentiation. Note, the $k$ terms help to transform the range of Jost's D to 0-1.

<br>
As a result, we expect that in populations that are distinct, Jost's D is close to **1**, as it means that alleles tend to be present in _a small number of individuals_ within the subpopulation (small $H_{Se}$). Individuals are “differentiated” in the sense that they contain different genetic material (large $H_{Te}$). 
<br>
Use the code chunk below to explore this statistics.  

```{r}
sim1$data %>%
  select(-locus) %>%
  group_by(time) %>%
  summarise(Js = mean(freq^2 + (1 - freq)^2),
            Jt = mean(freq)^2 + mean(1 - freq)^2,
            D = (Js - Jt) / Js * max(deme) / (max(deme) - 1)) %>%
  ggplot() + theme_bw() +
  geom_line(aes(x = time, y = D)) +
  expand_limits(y = c(0, 1.2))
```

Jost's D does not suffer from the same problem as $F_{ST}$, but it suffers from a different problem - it assumes infinite alleles mutation. When we have a biallelic SNPs, Jost's D will tend to reach values of around 0.5 at equilibrium.  

In conclusion, both $F_{ST}$ and D have their limitations. $F_{ST}$ is good when mutation rates are low, for example biallelic SNPs. Jost's D is good for highly diverse and fast-mutating sites, such as microsatellites.


### Connected subpopulations {-}
In this section, we will explore what happens when we connect our demes  by migration. In other words, we are relaxing our assumption of independent populations from above.  

_**Question/Coding Challenge**_: Run the simulator now with 1000 individuals, 5 demes, and a low migration rate. What do you observe in the allele frequencies? Try experimenting with larger and smaller migration rates in the following. What do you notice about the allele frequencies? 

```{r}
sim_freqs(N = 1000, demes = 5, mig_rate = 0.01)
```

#### Fst with Migration {-}
Lets look at what happens to $F_{ST}$ when we include migration. 
<br>
_**Question/Coding Challenge**_: What happens to the $F_{ST}$ values when you increase/decrease the migration rate? Does this make sense? 

```{r}
sim1 <- sim_freqs(N = 1000, demes = 10, t_out = seq(0, 365*20, 7), mig_rate = 1e-3,
                  mut_rate = 1e-6, plot_on = FALSE)
sim1$data %>%
  select(-locus) %>%
  group_by(time) %>%
  summarise(p_mean = mean(freq),
            p_var = popvar(freq),
            Fst = p_var / (p_mean * (1 - p_mean))) %>%
  ggplot() + theme_bw() +
  geom_line(aes(x = time, y = Fst)) +
  expand_limits(y = c(0, 1)) +
  geom_hline(yintercept = 1 / (1 + 2*1e3*(1e-3 + 1e-6)), linetype = "dashed")
```

Theory tells us that the equilibrium level of $F_{ST}$ in this situation is given by $\frac{1}{1 + 2N(m + mu)}$. This is a similar idea to the equilibrium we discussed above, but we are now balancing the forces of drift, migration, and mutation to reach a new equilibrium. 
<br>
_**Question/Coding Challenge**_: Does this equilibrium appear in your simulations? Note that migration rates are usually much larger than mutation rates, meaning we can often simplify the above equilibrium to $\frac{1}{1 + 2Nm}$.


Conclusion for $F_{ST}$ and migration: at equilibrium, $F_{ST}$ reaches a level that is a balance between the strength of genetic drift (i.e. small population size) and migration. The higher the migration rate, the lower the value of $F_{ST}$. This means we can use $F_{ST}$ to tell us something about the connectivity of populations. Small pairwise $F_{ST}$ tends to indicate highly connected populations and vice versa.

Note, however, that these patterns take a very long time to build up. So we are not talking about the timescale of control programmes, but more of an evolutionary timescale.

-------------------------

# Analysis of DRC data {-}

Now we will explore allele frequencies in a real dataset. First, we will load a dataset collected using Molecular Inversion Probes (MIPs) which are a method of high throughput sequencing. This dataset consists of 2537 samples collected in 2013–2015 from the DRC and surrounding countries. 
 
 This data is a slightly simplified version of that used in this paper : [Verity et al. 2020](https://pubmed.ncbi.nlm.nih.gov/32355199/).

```{r, echo=FALSE, eval=FALSE}
# Code to process raw data from DRC paper. This code block does not need to be run by the user.

# read in raw data
dat <- readRDS(url("https://github.com/bobverity/antimalarial_resistance_DRC/raw/master/source_data/biallelic_distances.rds"))

# simplify
dat$distance <- NULL
dat$samples <- dat$samples %>%
  select(ID, Country, ADM1NAME, Year, lat, long)
names(dat$samples) <- c("ID", "Country", "Admin1_name", "Year", "Latitude", "Longitude")

dat$loci <- dat$loci %>%
  select(CHROM_NUMERIC, POS, REF, ALT, NEUTRAL, GEO) %>%
  mutate(NEUTRAL = c("Neutral", "Non-neutral")[match(NEUTRAL, c("True", "False"))],
         GEO = c("Geographic", "Non-geographic")[match(GEO, c("True", "False"))])
names(dat$loci) <- c("CHROM", "POS", "REF", "ALT", "NEUTRAL", "GEO")

# get unique lat-lon into clusters
latlon <- interaction(round(dat$samples$Latitude, 2), round(dat$samples$Longitude, 2))
dat$samples$Cluster <- as.factor(match(latlon, unique(latlon)))

# save to file
saveRDS(dat, file = "data/DRC_MIPs_biallelic_processed.rds")
```

_**Coding Challenge**_: Have a look at the data using the *MIPanalyzer* package (*i.e.* just input the object to the console).

```{r}
# load data
MIP_data <- readRDS("data/DRC_MIPs_biallelic_processed.rds")

# have a look at the data
MIP_data

head(MIP_data$samples)

head(MIP_data$loci)
```

_**Coding Challenge**_: Now let's create a copy of the dataset that does not include any DRC samples using the ` MIPanalyzer::filter_samples` function:

```{r}
# filter samples
MIP_data_noDRC <- MIPanalyzer::filter_samples(MIP_data, MIP_data$samples$Country != "DRC")

MIP_data_noDRC
```


## PCA of DRC data {-}

Principal Components Analysis or PCA can be used to reduce the number of dimensions or variables in a dataset. As such, it is often used to more easily summarize data and in this case visualise genetic differentiation. 

_**Coding Challenge**_: Now using `MIPanalyzer::get_wsaf` and the `MIPanalyzer::pca_wsaf` function, let's create a PCA from the within-sample allele frequencies for samples not from the DRC. Then plot the resulting principal components from using the `geom_point` layer from ggplot. 

```{r}
wsaf_impute <- MIPanalyzer::get_wsaf(MIP_data_noDRC, impute = TRUE, FUN = mean)
pca <- MIPanalyzer::pca_wsaf(wsaf_impute)

plot_df <- data.frame(PC1 = pca$x[,1],
                      PC2 = pca$x[,2],
                      Country = MIP_data_noDRC$samples$Country)

ggplot(plot_df) + theme_bw() +
  geom_point(aes(x = PC1, y = PC2, color = Country))
```


We see well-separated clusters, which makes sense geographically. \

_**Coding Challenge**_: Now let's repeat the process using all samples including those from DRC. Use the code above to make this new PCA using the `MIP_data` dataframe (this code chunk is being hidden on purpose to let you practice!). 

```{r, echo=F}
wsaf_impute <- MIPanalyzer::get_wsaf(MIP_data, impute = TRUE, FUN = mean)
pca <- MIPanalyzer::pca_wsaf(wsaf_impute)

plot_df <- data.frame(PC1 = pca$x[,1],
                      PC2 = pca$x[,2],
                      Country = MIP_data$samples$Country)

ggplot(plot_df) + theme_bw() +
  geom_point(aes(x = PC1, y = PC2, color = Country))
```

The DRC points connect the other points. This suggests that there is more of a spatial continuum of allele frequencies in the DRC and that it may act as a "bridge" between the East and West African countries.

_**Question/Coding Challenge**_: Use the `MIPanalyzer::plot_pca_contribution` fucntion to plot the contributions that each locus makes to this PCA. Starting with PC1, what do you notice about the balance of geographically informative vs. non-informative SNPs?  What does this suggest?

```{r}
# plot component 1
MIPanalyzer::plot_pca_contribution(pca, component = 1, chrom = MIP_data$loci$CHROM, pos = MIP_data$loci$POS,
                                   locus_type = MIP_data$loci$GEO)
```
**_Answer_**: The pattern is mostly driven by geographically informative SNPs, which suggests that the primary factor driving this PCA is spatial location.

_**Question/Coding Challenge**_: As above, use the `MIPanalyzer::plot_pca_contribution` to repeat this process with PC2 (this code chunk is hidden on purpose to allow you to practice!). What do you notice about where these loci are on the genome? What genes are at these positions? You can use [PlasmoDB](https://plasmodb.org/plasmo/app/) to try and investigate these genomic locations. 

```{r, echo=F}
# plot component 2
MIPanalyzer::plot_pca_contribution(pca, component = 2, chrom = MIP_data$loci$CHROM, pos = MIP_data$loci$POS,
                                   locus_type = MIP_data$loci$GEO)
```
_**Answer**_: There all peaks at loci in one or two places. These are the pfcrt on chrom7 and then smaller peak at dhps on chromosome 8.

-------------------------

# Inferring population structure using *rmaverick* {-}

*rmaverick* is an R package which we will use to further investigate population structure. Please read the [package documentation](https://github.com/bobverity/rmaverick) to understand the various input parameters and model options.
<br>
<br>
First we will load genetic barcode data from [Bei et al., 2018](https://pubmed.ncbi.nlm.nih.gov/29325146/). This study contains samples taken at two separate time points over ten years apart in two locations in Senegal.  These data have already been filtered to remove the two Thies samples, and to order in terms of date (2001-2002 vs. 2014) and location (Dielmo and Ndiop). 

```{r, echo=FALSE, eval=FALSE}
# code not run - this is how I get the raw data from SIMPLEGEN and processed it
load("/Users/rverity/Dropbox/Bob/Work/My Programs/Simulation/SIMPLEGEN/data/Bei_2018.RData")
Bei_2018_processed <- Bei_2018

Bei_2018_processed$barcodes <- Bei_2018_processed$barcodes %>%
  filter(Location %in% c("Dielmo", "Ndiop")) %>%
  mutate(Date = as.Date(Date, format = "%m/%d/%Y"),
         Date_level = (Date > as.Date("2010-01-01")) + 1) %>%
  arrange(Date_level, Location, Date) %>%
  select(-Haplotype, -Date_level, -`Sample ID`, -COI) %>%
  as.data.frame()

Bei_2018_processed$EIR <- as.data.frame(Bei_2018_processed$EIR)

Bei_2018_processed$SNP_locations <- as.data.frame(Bei_2018_processed$SNP_locations)

save(Bei_2018_processed, file = "data/Bei_2018_processed.RData")
```

Now let's look at a snapshot of the data, which includes measures of transmission intensity in the two timepoints and regions. We will look at:
 - Entomological Innoculation Rate (EIR)
 - The barcodes
 - The SNP locations

```{r}
load("data/Bei_2018_processed.RData")

head(Bei_2018_processed$EIR)

head(Bei_2018_processed$barcodes)

head(Bei_2018_processed$SNP_locations)
```

Below, we are performing some data munging for `rmaverick`. Please copy and paste the code below into your console. 

```{r}
# filter out polygenomics
mav_data <- Bei_2018_processed$barcodes %>%
  filter(`M/P genomic` == "M")

# convert genotype calls to numeric
for (i in 8:31) {
  mav_data[,i] <- match(mav_data[,i], c("A", "C", "T", "G"))
}

head(mav_data)
```

Next, load the data into *rmaverick*, and set up our first model. Then print the project to check everything looks as planned (25 samples, 24 loci etc.)

```{r}
myproj <- rmaverick::mavproject() %>%
  rmaverick::bind_data(df = mav_data, ID_col = 1, data_cols = 8:31, ploidy = 1) %>%
  rmaverick::new_set(name = "no admixture model", admix_on = FALSE)

myproj
```

Markov Chain Monte Carlo (MCMC) approaches allow us to fit models by sampling from complex probability distributions. Here we will use MCMC to fit our model in rmaverick. 

Let's run an MCMC in rmaverick. Users should run this with pb_markdown = TRUE.

```{r}
myproj <- quiet(rmaverick::run_mcmc(myproj, K = 1:8, burnin = 1e3, samples = 1e3,
                              rungs = 10, GTI_pow = 1.5, pb_markdown =  FALSE))

```

First we will carry out some MCMC diagnostic checks. Check that coupling acceptance rates are above 0. Do this for every value of K (the number of subpopulations) from 2 to the max value explored.

_**Definition:** Coupling acceptance rates are the rates of proposals in the MCMC which are accepted._

_**Definition:** The K value is the number of subpopulations in rmaverick._


```{r}
rmaverick:::plot_mc_acceptance(myproj, K = 8)
```

_**Question/Coding Challenge**_: Using the `rmaverick::plot_logevidence_K` and `rmaverick::plot_posterior_K` functions, check what K values were found to have the most support. In this case, values around 3-5 are reasonably likely.

```{r}
rmaverick::plot_logevidence_K(myproj)
rmaverick::plot_posterior_K(myproj)
```

Finally,  we will produce a "STRUCTURE" plot. This plot shows differences in the distribution of genetic variants amongst populations  We will add vertical lines to break up the genetic data into the following groups:
1. Dielmo 2001-2001
2. Ndiop  2001-2002
1. Dielmo 2014
2. Ndiop  2014

_**Question/Coding Challenge**_: Using the `rmaverick::plot_qmatrix` fucntion to make the "STRUCTURE" plot and add x-intercepts at ` c(4, 10, 16) + 0.5` to break up the groups per above.
```{r}
rmaverick::plot_qmatrix(myproj, K = 3:5) +
  geom_vline(xintercept = c(4, 10, 16) + 0.5, size = 1)
```



Notice that the later samples are (mostly) clearly allocated to one or other population with high probability. These essentially represent clonal lineages. We can see that groups 2,3,4 are present in Dielmo in 2014, and in Ndiop we additionally have group 1.
<br>
Comparing this to the earlier time point, samples are far more ambiguous in their allocation. They look a bit like some of these clonal lineages, but it's not clear cut, probably because there is some additional variation. Interestingly, group 4 dominates in Ndiop in 2001-2002. This lineage is almost completely lost by the later sampling point.
<br>
Overall, we can see evidence of increasing population structure over time. Compare this with the EIR values - what do you notice? (lower transmission later in time, therefore increasing differentiation and population structure).

-------------------------
## Summary {-}
Concluding questions to discuss/summarize:  
1. What can allele frequencies tell us about population differentiation?  
2. What does $F_{ST}$ tell us and what are its limitations?  
3. When is population structure more ambiguous when using allele frequencies?  