---
title: "AMMS COI Practical" 
author: "Jason Hendry, Nick Brazeau"
date: "`r format(Sys.time(), '%B %d, %Y')`"
output: 
  html_document:
    theme: readable
    highlight: tango
    code_folding: hide
    toc: yes
    toc_float:
      collapsed: false
      smooth_scroll: false
    toc_depth: 2
    number_sections: true
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = T, warning = F, message = F, 
                      fig.align = 'center', fig.keep = 'all')
```

<!-- Here we style out button a little bit -->

```{=html}
<style>
.showopt {
background-color: #004c93;
color: #FFFFFF; 
width: 100px;
height: 20px;
text-align: center;
vertical-align: middle !important;
border-radius: 8px;
float:right;
}

.showopt:hover {
background-color: #dfe4f2;
color: #004c93;
}

</style>
```
<!--Include script for hiding output chunks-->

```{=html}
<script src="misc/hideOutput.js"></script>
```
# Dependencies for Practical {.unnumbered}

Please copy and paste the below code chunk in it's entirety to your console to download R package libraries needed for this practical. If you are having trouble installing any of the R packages, please ask an instructor for a pre-loaded flash drive.

```{r, echo=T}
deps <- c("tidyverse", "vcfR", "McCOILR")
deps <- !sapply(deps, function(x){x %in% installed.packages()[,1]} )
if(any(deps)) {
  if(deps["McCOILR"]) {
    if (!"remotes" %in% installed.packages()[,1]){
      install.packages("remotes")
    }
    remotes::install_github("OJWatson/McCOILR")
    deps <- deps[names(deps) != "McCOILR"]
  } else {
    install.packages(names(deps)[deps])
  }
}


```

Please now load all of those libraries into this session using the code chunk below. Please copy and paste it in its entirety.

```{r, echo=T}
library(tidyverse)
library(vcfR)
library(McCOILR)
```

Please set *your* working director so that you are in the `AMMS/COI` directory. If you are having trouble finding that path, you can use the `file.choose` function to help determine its location. 

# Practical Goals {.unnumbered}

By the end of this practical, you should understand the following concepts:  
- ***

------------------------------------------------------------------------

# Importing Data

Here, we will use the `vcfR` [package](https://knausb.github.io/vcfR_documentation/) to read in our variant call file (VCF) that contains biallelic SNPs for our *** individuals across our four populations.

```{r}
DRCongo <- vcfR::read.vcfR("data/SNPs.DRCongo.setA.random96.vcf")
#Vietnam <- vcfR::read.vcfR("data/SNPs.Vietnam.setD.random96.vcf")
```

## Munging & Extracting Genomic Data for DRC
*** 
- going to go from wide to long for ease 
```{r}
# extract information
# loci <- vcfR::extract_info_tidy(DRCongo)
#TODO Jason *** - above throws an error, it shouldn't unless vcf is misbehaving. haven't dug in. longer approach below

loci <- DRCongo@fix[,1:2] %>% 
  tibble::as_tibble() %>% 
  dplyr::mutate(POS = as.numeric(POS),
                Key = 1:dplyr::n()) %>% 
  dplyr::select(c("CHROM", "POS", "Key"))


# extract GT information and get WSRAF for every sample at each loci
DRCongo_long <- DRCongo %>% 
  vcfR::extract_gt_tidy() %>% 
  # lets make some new variables
  dplyr::mutate(
    # need to change the GT character to a numeric for maths
    gt = dplyr::case_when(gt_GT == "0/0" ~ 0,
                          gt_GT == "0/1" ~ 0.5,
                          gt_GT == "1/1" ~ 1), 
    # get reference allele depth
    rad = purrr::map_dbl(gt_AD, function(x){vcfR::masplit(
      as.matrix(x), record = 1, sort = FALSE, decreasing = FALSE)}),
    # get alternate allele depth 
    aad = purrr::map_dbl(gt_AD, function(x){vcfR::masplit(
      as.matrix(x), record = 2, sort = FALSE, decreasing = FALSE)}),
    # calculate within-sample reference allele freq
    wsraf = rad/(rad + aad),
    wsraf = ifelse(is.nan(wsraf), NA, wsraf) # occurs when 0/0
    ) %>% 
  # now let's select the variables that we want
  dplyr::select(c("Key", "Indiv", "gt", "wsraf")) %>% 
  # now lets combine the loci information with the individual level information
  dplyr::full_join(x = loci, y = ., by  = "Key") %>% 
  # don't need Key anymore
  dplyr::select(-c("Key"))

```

### Look at VCF in "long" format 
Each of these columns *** 
```{r}
head(DRCongo_long)
```

# Population Level Allele Frequencies
Now we will calculate the PLAF by ***

```{r}
DRC_plaf <- DRCongo_long %>% 
  dplyr::group_by(CHROM, POS) %>% 
  dplyr::summarise(
    # multiple by 2 to assume diploid genome
    PLAF = sum(gt * 2, na.rm = T) / (2* sum(!is.na(gt))) 
  )

```

### Visualizing PLAF

```{r}
DRC_plaf %>% 
  ggplot() + 
  geom_histogram(aes(x = PLAF))
```

# Calculating COI
***

## Reformatting data for RMCL
***
```{r}
# wide format 
DRCongo_RMCL <- DRCongo_long %>% 
  dplyr::mutate(loci = paste(CHROM, POS, sep = "|")) %>% 
  dplyr::select(c("loci", "Indiv", "gt")) %>% 
  # liftover missing for RMCL 
  dplyr::mutate(gt = ifelse(is.na(gt), -1, gt)) %>% 
  tidyr::pivot_wider(names_from = "Indiv",
                     values_from = "gt")
# make matrix and format for RMCL
DRCongo_RMCLmat <- as.matrix(DRCongo_RMCL[2:ncol(as.matrix(DRCongo_RMCL))])
rownames(DRCongo_RMCLmat) <- DRCongo_RMCL[["loci"]]
DRCongo_RMCLmat <- t(DRCongo_RMCLmat)
# make list of multiple SNP sets
gts_rmcl <- list("DRCsetA" = DRCongo_RMCLmat)

```

## Running RMCL
***
Do this into the /COI directory
`git clone https://github.com/EPPIcenter/THEREALMcCOIL.git`
Also see: https://github.com/EPPIcenter/THEREALMcCOIL/tree/master/categorical_method

```{r}
# Prepare output directory
output_dir <- "results"
if (!dir.exists(output_dir)) {
  dir.create(output_dir)
}
out_cat <- McCOIL_categorical(gts_rmcl[["DRCsetA"]],
                              maxCOI=25, threshold_ind=20,
                              threshold_site=20,
                              totalrun=1000, burnin=100, M0=15,
                              e1=0.05, e2=0.05,
                              err_method=3, 
                              path="results",
                              output="output_DRCsetA.txt" )


```

## Exploring THEREALMcCOIL outputs
***
```{r}
# Load summary data
DRCongo_COI <- readr::read_table("results/output_DRCsetA.txt_summary.txt") %>% 
  # Subset to COI results
  dplyr::filter(CorP == "C") %>% 
  # select to rows we care about
  dplyr::select(-c("file", "CorP")) %>% 
  dplyr::rename(Indiv = name)

# Compute fraction mixed
DRCongo_COI %>% 
  dplyr::summarise(frac_mixed = sum(mean > 1) / dplyr::n()) 
  
# What does the distribution look like?
DRCongo_COI %>% 
  ggplot() + 
  geom_histogram(aes(x = mean))
  


```
## TODO think of more exploration 

# Heterozygosityand Fws
## Exploring WSAF
***
```{r}
# pick five individuals
fivefriends <- DRCongo_long$Indiv[1:5]
DRCongo_long %>% 
  dplyr::filter(Indiv %in% fivefriends) %>% 
  ggplot() + 
  geom_histogram(aes(x = wsraf)) +
  facet_wrap(~Indiv)

```
### TODO - have participants explore wsraf vs COI call? 


### Comparing PLAF vs mean WSAFs
- Linear -- not at all surprising
- Both PLAF and mean WSAF per-site driven by # samples not REF
***

```{r}

DRCongo_long %>% 
  dplyr::group_by(CHROM, POS) %>% 
  dplyr::summarise(meanWSRAF = mean(wsraf, rm.na = T)) %>% 
  dplyr::full_join(DRC_plaf, ., by = c("CHROM", "POS")) %>% 
  ggplot() + 
  geom_point(aes(x = PLAF, y = meanWSRAF))

```

### Compute heterozygosities
***
```{r}
calc_biallelic_heterozygosity <- function(p_alt) {
  return (p_alt * (1 - p_alt))
}

# calculate het for samples
DRCongo_long <- DRCongo_long %>% 
  dplyr::mutate(
    het_wsafs = purrr::map_dbl(wsraf, calc_biallelic_heterozygosity)
  )

# calculate het for pop
DRC_plaf <- DRC_plaf %>% 
  dplyr::mutate(
    het_plaf = purrr::map_dbl(PLAF, calc_biallelic_heterozygosity)
  )

# now calculate fws
DRCongo_fws <- dplyr::full_join(DRCongo_long, DRC_plaf, by = "CHROM", "POS") %>% 
  dplyr::group_by(Indiv) %>% 
  dplyr::summarise(
    fws = mean( 1 - het_wsafs / het_plaf, na.rm = T)
  )


```

### Visualize Heterozygosity
***
```{r}
DRCongo_fws %>% 
  ggplot() + 
  geom_histogram(aes(x = fws))

```

# COI vs Heterozygosity Relationships
***
```{r}
# now combine for a final dataframe
DRCongo_final <- DRCongo_long %>% 
  dplyr::left_join(., DRCongo_COI, by = "Indiv") %>% 
  dplyr::left_join(., DRCongo_fws, by = "Indiv")

# plot
DRCongo_final %>% 
  dplyr::mutate(mean_char = as.factor(mean)) %>% 
  ggplot() + 
  geom_boxplot(aes(x = mean_char, y = fws))

```
#### TODO more relationships
